packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +104,224 @@ if (process.env.NODE_ENV === 'development') {
   OverflowContainerContext.displayName = 'OverflowContainerContext';
 }
 
+/**
+ * Hook that returns the number of items that should be visible based on the size of the container element.
+ *
+ * The returned number should be used to render the element with fewer items.
+ *
+ * @private
+ * @param items Items that this element contains.
+ * @param dimension 'horizontal' (default) or 'vertical'
+ * @returns [callback ref to set on container, stateful count of visible items]
+ *
+ * @example
+ * const items = Array(10).fill().map((_, i) => <span>Item {i}</span>);
+ * const [ref, visibleCount] = useOverflow(items);
+ * ...
+ * return (
+ *   <div ref={ref}>
+ *     {items.slice(0, visibleCount)}
+ *   </div>
+ * );
+ */
+const useOverflow = <T extends HTMLElement>(
+  itemsCount: number,
+  orientation: 'horizontal' | 'vertical' = 'horizontal',
+) => {
+  /** `[number, number]` means that we're still guessing. `null` means that we got the correct `visibleCount`. */
+  type GuessRange = [number, number] | null;
+
+  /** First guess of the number of items that overflows. We refine this guess with subsequent renders. */
+  const STARTING_MAX_ITEMS_COUNT = 32;
+
+  const containerRef = React.useRef<T>(null);
+
+  const initialVisibleCount = React.useMemo(
+    () => Math.min(itemsCount, STARTING_MAX_ITEMS_COUNT),
+    [itemsCount],
+  );
+  const initialVisibleCountGuessRange = React.useMemo(
+    () => [0, initialVisibleCount] satisfies GuessRange,
+    [initialVisibleCount],
+  );
+
+  const [visibleCount, _setVisibleCount] =
+    React.useState<number>(initialVisibleCount);
+  const [visibleCountGuessRange, setVisibleCountGuessRange] =
+    React.useState<GuessRange>(initialVisibleCountGuessRange);
```

Comment:
This whole "range" idea feels confusing and even misleading to me. The first thing I would suggest is to use an object instead of array, so that we are not dealing with indexes.

---

Further, I think we can reduce the number of different states (while keeping the same algorithm).

There are four (three and a half?) different states right now:
1. min guess
2. max guess
3. `visibleCount`
4. `isStabilized` indicated by the "range" being `null`.

The "max guess" makes sense, but the "min guess" and `null` parts are what feel unnecessary. The "min guess" is basically the "stale" version of the "max guess", and also often overlaps with the `visibleCount`. And setting `null` also feels unnecessary — we could probably derive it from other existing states.

Reply:
The reducer does make things cleaner, so I like that. I need some more time to think about the other points before I respond.

In the meantime, can you investigate the failing e2e tests?

Reply:
After all the recent changes, I feel good about this. The reducer makes things a lot easier to understand, since it abstracts multiple states into well-named actions. 👍